core/src/avm2/globals/flash/display/graphics.rs

```diff
@@ -41,13 +43,93 @@ pub fn begin_fill<'gc>(
 /// Implements `Graphics.beginBitmapFill`.
 pub fn begin_bitmap_fill<'gc>(
     activation: &mut Activation<'_, 'gc>,
-    _this: Option<Object<'gc>>,
-    _args: &[Value<'gc>],
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
 ) -> Result<Value<'gc>, Error<'gc>> {
-    avm2_stub_method!(activation, "flash.display.Graphics", "beginBitmapFill");
+    if let Some(this) = this.and_then(|t| t.as_display_object()) {
+        let bitmap = args
+            .get_object(activation, 0, "bitmap")?
+            .as_bitmap_data()
+            .expect("Bitmap argument is ensured to be a BitmapData from actionscript");
+        let matrix = if let Some(matrix) = args.try_get_object(activation, 1) {
+            swf_matrix_from_object(activation, matrix)?
+        } else {
+            // Users can explicitly pass in `null` to mean identity matrix
+            Matrix::IDENTITY
+        };
+        let is_repeating = args.get_bool(2);
+        let is_smoothed = args.get_bool(3);
+
+        let handle = if let Some(handle) = bitmap
+            .write(activation.context.gc_context)
+            .bitmap_handle(activation.context.renderer)
+        {
+            handle
+        } else {
+            return Ok(Value::Undefined);
+        };
+
+        let bitmap = ruffle_render::bitmap::BitmapInfo {
+            handle,
+            width: bitmap.read().width() as u16,
+            height: bitmap.read().height() as u16,
+        };
+        let scale_matrix = Matrix::scale(
+            Fixed16::from_f64(bitmap.width as f64),
+            Fixed16::from_f64(bitmap.height as f64),
+        );
+
+        if let Some(mut draw) = this.as_drawing(activation.context.gc_context) {
+            let id = draw.add_bitmap(bitmap);
+            draw.set_fill_style(Some(FillStyle::Bitmap {
+                id,
+                matrix: matrix * scale_matrix,
+                is_smoothed,
+                is_repeating,
+            }));
+        }
+    }
+
     Ok(Value::Undefined)
 }
 
+fn swf_matrix_from_object<'gc>(
```

Comment:
could we reuse `fn object_to_matrix` here, at least to grab the AS fields?
(kinda weird that we have two different matrix types here, but I'm not questioning which one is right-er in which scenario)

Reply:
Looks like we already added `From<ruffle_render::matrix::Matrix> for swf::Matrix`, so I've made that change!

swf Matrix is lossy and render one isn't... otherwise I don't know. It was a change made a very long time ago to split them up.